### PR TITLE
Basic Search patient by address, gender, phone

### PR DESCRIPTION
**Pull request by silvistershebin**
#### Short description of what this resolves:
Search patient by name, dob, fname, lname was implemented in the getAll($search) function in PatientService.php. However while calling this function from the FhirPatientRestController, only the dob and name parameters were passed. Due to this, search by fname, lname were not working.

#### Changes proposed in this pull request:
Added the lname and fname parameters in FhirPatientRestController.

### CODE REVIEW

File: src/RestControllers/FhirPatientRestController.php [hunk 1]
```diff
@@ -78,7 +78,9 @@ public function getAll($search)
 
         $searchParam = array(
             'name' => $search['name'],
-            'dob' => $search['birthdate']
+            'dob' => $search['birthdate'],
```
Comment (bradymiller) [inline src/RestControllers/FhirPatientRestController.php]: to support $search['birthdate'], looks like also need to incorporate that in conditional at line 227 in /src/Services/PatientService.php

File: src/RestControllers/FhirPatientRestController.php [hunk 2]
```diff
@@ -78,7 +78,9 @@ public function getAll($search)
...
+			'fname' => $search['fname'],
+			'lname' => $search['lname']
```
Comment (bradymiller) [inline src/RestControllers/FhirPatientRestController.php]: Is there a standard terminology for what can be searched for via FHIR?

Comment (silvistershebin) [reply src/RestControllers/FhirPatientRestController.php]: Yes there is a set of terminologies: https://www.hl7.org/fhir/searchparameter-registry.html
I think fname and lname are not mentioned in this set. However the implementation for fname and lname was done in PatientService.php and thats why I added it in FhirPatientRestController.php.

File: src/RestControllers/FhirPatientRestController.php (hunk 1 above)
Comment (silvistershebin) [reply src/RestControllers/FhirPatientRestController.php]: In FhirPatientRestController.php, the $search['birthdate'] is assigned to 'dob' in the $searchParam array. 'dob' has been included in the 'if' condition on line 227 of PatientService.php.

File: src/RestControllers/FhirPatientRestController.php (hunk 2 above)
Comment (silvistershebin) [reply src/RestControllers/FhirPatientRestController.php]: I have my code ready for search abiding by the given terminologies. I can push that code and revert the fname and lname if required.

File: src/RestControllers/FhirPatientRestController.php (hunk 1 above)
Comment (bradymiller) [reply src/RestControllers/FhirPatientRestController.php]: oops, thanks for pointing that out

File: src/RestControllers/FhirPatientRestController.php (hunk 2 above)
Comment (bradymiller) [reply src/RestControllers/FhirPatientRestController.php]: I suppose it depends on how strict we need to be. If we need to be strict, then could basically set fname and lname to also equal $search['name'] (so sending search for name would work as full name, last name, or first name). Will let FHIR superhero @sjpadgett weigh in in this one.

Comment (silvistershebin) [reply src/RestControllers/FhirPatientRestController.php]: Ok sure.

**Comment by sjpadgett** [comment]
A couple things. Definitely need to follow Fhir search queries.  However, we need to restrict what queries we will accept. This is done through a capability statement but for now we stick to basic searches.
So any search query for a resource will expect a search by the fhir resource property ie name. Because we don't have repositories where we store fhir objects we have to translate to something our standard services can understand. Thus we need an adapter. So any search would run through the adapter to do these translations. eg name can be any value of the name so our lookup might concat fname+lname+mname then compare etc.

I'm not sure if we want to build the resource search to each controller rather than calling an adapter to do the job but for now, we'll leave it as here.
Currently i'm trying to get our test jig working again in 5.0.3 where you can talk to a public HAPI server for testing or call openemr fhir api(with auth) to test our stuff both ways. I'm almost there but I can't keep up with you guys.
image

This is result of all encounters for patient returned as a bundled resource from HAPI server.
$query = [
            'patient' => $id,
            '_format' => 'json',
            '_pretty' => 'true'
        ];
        $response = oeHttp::usingBaseUri($this->fhirBase)->get('Encounter', $query);

@silvistershebin My suggestion to you for now is to use the HAPI public server to test your queries. That way you'll know what will get returned from a full fhir server.
Just point to http://hapi.fhir.org/baseR4 where you can create patients or do general searches or go to http://hapi.fhir.org/ and there are crud interfaces with what others have created. I'd go there first and it will open your eyes quickly. Also I have two test patients on that server namely patient-30 and patient-46 you can use to query against.

I hope to have the Publish test jig in codebase today and I hope i've made some sense here.

File: src/RestControllers/FhirPatientRestController.php (hunk 2 above)
Comment (sjpadgett) [reply src/RestControllers/FhirPatientRestController.php]: > I have my code ready for search abiding by the given terminologies. I can push that code and revert the fname and lname if required.

Our rest services are setup for rest queries so yes revert.

**Comment by sjpadgett** [comment]
@silvistershebin 
You need to fix travis errors. Brady uses this
~/git/PHP_CodeSniffer/bin/phpcbf -n --extensions=php,inc --standard=ci/phpcs.xml
but I don't know what your dev setup is so, may as well get making travis happy out of the way:)

**Comment by bradymiller** [comment]
Hi @silvistershebin , To avoid us causing  "travis post traumatic stress disorder" (t-PTSD), I'll bring this in and do a quick PSR2 run on it. Soon the easy development docker (thanks to @prondubuisi) will support doing PSR2 fixes, which will drastically improve the quality of life of OpenEMR developers around the world :)

**Pull request merged**